
```{r libraries}
library(tidyverse)
library(janitor)
``` 

```{r read in files from command line}
#args <- commandArgs(trailingOnly = TRUE)
#snps <- read.csv(args[1],sep='\t',header=T) 
#low_cov <- read.csv(args[2],sep='\t',header=T)
#is_short <- read.csv(args[4],sep='\t',header=T) 
```

```{r format snp csv}
snps <- read.csv("14_enterocinA_variation/summary/enterocinA_summary.csv", sep=';', header=T,row.names=NULL) 
colnames(snps) <- colnames(snps)[2:ncol(snps)]
snps <- snps[1:(ncol(snps)-1)]
samples <- snps %>% select(-c(CHROM,POS,ID,REF,ALT,QUAL,FILTER,FORMAT)) %>% colnames()

snps_pivot <- pivot_longer(snps,cols=all_of(samples), names_to= "Clone", values_to="Qual_metrics")
snps_pivot$Clone <- gsub("\\.", "-", snps_pivot$Clone)

#separate quality metric values into their own columns
snps_pivot1 <- snps_pivot %>% separate(Qual_metrics,c('DP','RO','AO','QA'),sep =':')

#fix formating on QA and AO values that were formatted incorrectly during vcf merge
snps_pivot1$QA <- gsub("\\[|\\]", "", snps_pivot1$QA)
snps_pivot1$QA <- gsub("\\None, ", "", snps_pivot1$QA)
snps_pivot1$QA <- gsub("\\, None", "", snps_pivot1$QA)
snps_pivot1$AO <- gsub("\\[|\\]", "", snps_pivot1$AO)
snps_pivot1$AO <- gsub("\\None, ", "", snps_pivot1$AO)
snps_pivot1$AO <- gsub("\\, None", "", snps_pivot1$AO)

#keep only rows where clone has the snp
snps_filter <- filter(snps_pivot1, DP != "None") 
snps_filter1 <- snps_filter %>% select(-c(CHROM,ID,FILTER,FORMAT,QUAL)) #remove unnecessary columns

snps_filter1$AO <- as.integer(snps_filter1$AO)

#filter positions to 246177 - 256268
snps_filter2 <- snps_filter1
snps_filter2$between <- between(snps_filter2$POS,246177,256268)
snps_filter2<- filter(snps_filter2,between ==TRUE) %>% select(-c("between"))

```

```{r get snp annotations}
ann_snps <- read.csv("17_snpeff/enterocinA/summary/enterocinA_annotated_summary.csv", sep=';', header=T,row.names=NULL) 
colnames(ann_snps) <- colnames(ann_snps)[2:ncol(ann_snps)]
ann_snps <- ann_snps[1:(ncol(ann_snps)-1)]

ann_snps <- ann_snps %>% select(POS,REF,ALT,X.Annotation.Annotation_Impact.Gene_Name.Gene_ID.Feature_Type.)

#filter positions to 246177 - 256268
ann_snps$between <- between(ann_snps$POS,246177,256268)
ann_snps<- filter(ann_snps,between ==TRUE) %>% select(-c("between"))
```


```{r format low cov regions-10% enterocin A region only}
low_covA10 <- read.csv("14_enterocinA_variation/summary/enterocinA_summary.low_coverage_intervals_enterocinA10.txt",sep='\t',header=T) 
low_cov_filtA10 <- low_covA10 %>% select(-c(Chromosome))
low_cov_format10 <- unite(low_cov_filtA10, col = "Low_cov_region", c(Start_Position, Stop_Position), sep="-")
low_cov_table10 <- low_cov_format10  %>% select(Low_cov_region) %>% table() %>% as.data.frame() 


##FILTER low cov positions after manual inspection 
#bp between 249122 and 249141 is not real and should be included in the low cov region
#bp between 251770 and 251789 is not real and should be included in the low cov region
#bp between 256290 and 256350 is not real and should be included in the low cov region

low_cov_filtA10_order <- low_cov_filtA10[order(low_cov_filtA10$Clone),] 

dataframe_rows_to_remove <- data.frame()
i=1
while (i < nrow(low_cov_filtA10_order)){
    if (low_cov_filtA10_order$Clone[i] == low_cov_filtA10_order$Clone[i+1]) {
        if (low_cov_filtA10_order$Stop_Position[i]==249122 && low_cov_filtA10_order$Start_Position[i+1]==249141){
            low_cov_filtA10_order$Stop_Position[i] <- low_cov_filtA10_order$Stop_Position[i+1]
            dataframe_rows_to_remove  <- dataframe_rows_to_remove %>% rbind(low_cov_filtA10_order[i+1,])
        }
    }
    i=i+1
}   
low_cov_filtA10_order2 <- anti_join(low_cov_filtA10_order, dataframe_rows_to_remove, by = c("Clone","Start_Position","Stop_Position"))


dataframe_rows_to_remove <- data.frame()
y=1
while (y < nrow(low_cov_filtA10_order2)){
    if (low_cov_filtA10_order2$Clone[y] == low_cov_filtA10_order2$Clone[y+1]) {
        if (low_cov_filtA10_order2$Stop_Position[y]==251770 && low_cov_filtA10_order2$Start_Position[y+1]==251789){
            low_cov_filtA10_order2$Stop_Position[y] <- low_cov_filtA10_order2$Stop_Position[y+1]
            dataframe_rows_to_remove  <- dataframe_rows_to_remove %>% rbind(low_cov_filtA10_order2[y+1,])
        }
    }
    y=y+1
}  
low_cov_filtA10_order3 <- anti_join(low_cov_filtA10_order2, dataframe_rows_to_remove, by = c("Clone","Start_Position","Stop_Position"))



dataframe_rows_to_remove <- data.frame()
z=1
while (z < nrow(low_cov_filtA10_order3)){
    if (low_cov_filtA10_order3$Clone[z] == low_cov_filtA10_order3$Clone[z+1]) {
        if (low_cov_filtA10_order3$Stop_Position[z]>256290 && low_cov_filtA10_order3$Start_Position[z+1]<256351){
            low_cov_filtA10_order3$Stop_Position[z] <- low_cov_filtA10_order3$Stop_Position[z+1]
            dataframe_rows_to_remove  <- dataframe_rows_to_remove %>% rbind(low_cov_filtA10_order3[z+1,])
        }
    }

    z=z+1
}  
low_cov_filtA10_order4 <- anti_join(low_cov_filtA10_order3, dataframe_rows_to_remove, by = c("Clone","Start_Position","Stop_Position"))


#filter positions to 246177 - 256268
low_cov_filtA10_order5 <- low_cov_filtA10_order4
low_cov_filtA10_order5$st_btwn_st <- between(low_cov_filtA10_order5$Start_Position, 245800,246177)
low_cov_filtA10_order5$end_btwn_st <- between(low_cov_filtA10_order5$Stop_Position, 245800,246177)
low_cov_filtA10_order5$st_btwn_end<- between(low_cov_filtA10_order5$Start_Position, 256268,256750)
low_cov_filtA10_order5$end_btwn_end<- between(low_cov_filtA10_order5$Stop_Position, 256268,256750)


low_cov_1 <- low_cov_filtA10_order5 
#if end_btwn_end==T --> set low cov end position to 256268
low_cov_1$Stop_Position[low_cov_1$end_btwn_end==TRUE & low_cov_1$st_btwn_st==FALSE & low_cov_1$end_btwn_st==FALSE & low_cov_1$st_btwn_end==FALSE] <- 256268

#if st_btwn_st ==T --> set low cov start position to 246177
low_cov_1$Start_Position[low_cov_1$end_btwn_end==FALSE & low_cov_1$st_btwn_st==TRUE & low_cov_1$end_btwn_st==FALSE & low_cov_1$st_btwn_end==FALSE] <- 246177

#if st_btwn_st and end_btwn_end ==T --> low cov region = 246177 - 256268 (isolate will be removed from consideration)
low_cov_1$Start_Position[low_cov_1$end_btwn_end==TRUE & low_cov_1$st_btwn_st==TRUE & low_cov_1$end_btwn_st==FALSE & low_cov_1$st_btwn_end==FALSE] <- 246177
low_cov_1$Stop_Position[low_cov_1$end_btwn_end==TRUE & low_cov_1$st_btwn_st==TRUE & low_cov_1$end_btwn_st==FALSE & low_cov_1$st_btwn_end==FALSE] <- 256268

#if st_btwn_end and end_btwn_end ==T --> remove low cov region
#if st_btwn_st and end_btwn_st ==T --> remove low cov region
low_cov_2 <- low_cov_1 %>% filter(!(st_btwn_end==TRUE & end_btwn_end==TRUE))
low_cov_3 <- low_cov_2 %>% filter(!(st_btwn_st==TRUE & end_btwn_st==TRUE))

low_cov_3 <- low_cov_3 %>% select(c(Clone,Start_Position,Stop_Position))
colnames(low_cov_3) <- c("Clone","Low_cov_start_pos","Low_cov_end_pos")

low_cov_format10 <- unite(low_cov_3, col = "Low_cov_region", c(Low_cov_start_pos, Low_cov_end_pos), sep="-")
low_cov_table10 <- low_cov_format10  %>% select(Low_cov_region) %>% table() %>% as.data.frame() 

```



```{r remove snps that fall in low cov regions - 10% low cov}
snps_lcA10 <- merge(snps_filter2,low_cov_3, by="Clone", all=TRUE)
snps_lcA10$between <- between(snps_lcA10$POS,snps_lcA10$Low_cov_start_pos,snps_lcA10$Low_cov_end_pos)
snps_lcA10$AO <- as.integer(snps_lcA10$AO)

snps_lc_btwn_trueA10<- filter(snps_lcA10,between ==TRUE)

#remove snp calls that fall in low cov regions 
merged_snps_lc <- merge(snps_filter2,snps_lc_btwn_trueA10, by=c("POS","Clone","REF","ALT","DP","RO","AO","QA"), all.x=TRUE )
accepted_variants <- merged_snps_lc [is.na(merged_snps_lc $between), ] #only keep snp calls where the call in not in a low cov region
num_snps <- accepted_variants %>% select(c(POS,REF,ALT)) %>% distinct()

#exlcude manually inspected snp calls that shoud be excluded
filter_positions <- c(247121,249121,256296,256313,256335,256342,247637,250326,250429,250717,251762)
accepted_variants_filter1 <- accepted_variants %>% filter(!POS %in%filter_positions )
accepted_variants_filter2 <- accepted_variants_filter1 %>% filter(!(POS==251950 & Clone =="BL03349-1")) %>% 
                                                        filter(!(POS==248521 & Clone =="BL03200-1")) %>% 
                                                        filter(!(POS==248546 & Clone =="BL03200-1")) %>% 
                                                        filter(!(POS==249543 & Clone =="BL03349-1"))



final_num_snps <- accepted_variants_filter2 %>% select(c(POS,REF,ALT)) %>% distinct()


accepted_variants_filter2 <- accepted_variants_filter2 %>% select(c(POS,Clone,REF,ALT))

#add column with variant annotation
annotated_accepted_variants <- merge(accepted_variants_filter2, ann_snps, all.x=TRUE)
```

```{r merge accepted variant list with low cov list - 10% low cov}
#pivot variant list so it can be merged to make a master list
variant_pivot <- pivot_wider(annotated_accepted_variants, id_cols="Clone", names_from= "POS", values_from="X.Annotation.Annotation_Impact.Gene_Name.Gene_ID.Feature_Type.")

#format and pivot low cov region file for merge
low_cov_format <- unite(low_cov_3, col = "Low_cov_region", c(Low_cov_start_pos, Low_cov_end_pos), sep="-")
low_cov_table <- low_cov_format  %>% select(Low_cov_region) %>% table() %>% as.data.frame() #count number of unique low cov regions
low_cov_pivot <- pivot_wider(low_cov_format, id_cols="Clone", names_from= "Low_cov_region", values_from="Low_cov_region")

variant_low_cov <- merge(variant_pivot,low_cov_pivot,by="Clone",all=TRUE)


```




```{r read in IS elements and merge with variants and low cov}
#is_short <- read.csv("14_enterocinA_variation/summary/enterocinA_summary.ISFinder.txt",sep='\t',header=T) 
is_long <- read.csv("14_enterocinA_variation/summary/enterocinA_summary.long.ISFinder.txt",sep='\t',header=T) 

#filter positions to 246177 - 256268
is_long$between <- between(is_long$Start_Position,246177,256268)
is_long<- filter(is_long,between ==TRUE) %>% select(-c("between"))

is_short <- is_long %>%
    select(-c(Sample,Chromosome, Stop_Position, Potential_sequence, Alignment)) %>%
    filter(Potential_IS != "No identity") %>%
    pivot_wider(names_from = Potential_IS, values_from = Start_Position, values_fn = list)

#separate split IS values
is_short$IS1678 <- as.character(is_short$IS1678)
is_short$ISLgar5 <- as.character(is_short$ISLgar5)

islgar_1 <- is_short %>% filter(ISLgar5 == "c(246648, 248570)" )
islgar_1_split <- as.data.frame(str_extract(string = islgar_1$ISLgar5,pattern = "(?<=\\().*(?=\\))") )
islgar_1_values <- as.data.frame(unlist(strsplit(islgar_1_split[1,1],','))) %>% t()
rownames(islgar_1_values) <- NULL
colnames(islgar_1_values) <- c("ISLgar5","ISLgar5_2")
islgar_1_edit <- islgar_1 %>% select(-c("ISLgar5"))
islgar_1_edit1 <- cbind(islgar_1_edit,islgar_1_values)

islgar_2 <- is_short %>% filter(ISLgar5 == "c(246293, 246648)" )
islgar_2_split <- as.data.frame(str_extract(string = islgar_2$ISLgar5,pattern = "(?<=\\().*(?=\\))") )
islgar_2_values <- as.data.frame(unlist(strsplit(islgar_2_split[1,1],','))) %>% t()
rownames(islgar_2_values) <- NULL
colnames(islgar_2_values) <- c("ISLgar5_2","ISLgar5")
islgar_2_edit <- islgar_2 %>% select(-c("ISLgar5"))
islgar_2_edit1 <- cbind(islgar_2_edit,islgar_2_values)

islgar_3 <- is_short %>% filter(ISLgar5 == "c(246211, 246648)" )
islgar_3_split <- as.data.frame(str_extract(string = islgar_3$ISLgar5,pattern = "(?<=\\().*(?=\\))") )
islgar_3_values <- as.data.frame(unlist(strsplit(islgar_3_split[1,1],','))) %>% t()
rownames(islgar_3_values) <- NULL
colnames(islgar_3_values) <- c("ISLgar5_2","ISLgar5")
islgar_3_edit <- islgar_3 %>% select(-c("ISLgar5"))
islgar_3_edit1 <- cbind(islgar_3_edit,islgar_3_values)

#
is1678_1 <- is_short %>% filter(IS1678 == "c(248570, 250498)" )
is1678_1_split <- as.data.frame(str_extract(string = is1678_1$IS1678,pattern = "(?<=\\().*(?=\\))") )
is1678_1_values <- as.data.frame(unlist(strsplit(is1678_1_split[1,1],','))) %>% t()
rownames(is1678_1_values) <- NULL
colnames(is1678_1_values) <- c("IS1678","IS1678_2")
is1678_1_edit <- is1678_1 %>% select(-c("IS1678"))
is1678_1_edit1 <- cbind(is1678_1_edit,is1678_1_values)

is1678_2 <- is_short %>% filter(IS1678 == "c(248036, 248570)" )
is1678_2_split <- as.data.frame(str_extract(string = is1678_2$IS1678,pattern = "(?<=\\().*(?=\\))") )
is1678_2_values <- as.data.frame(unlist(strsplit(is1678_2_split[1,1],','))) %>% t()
rownames(is1678_2_values) <- NULL
colnames(is1678_2_values) <- c("IS1678_2","IS1678")
is1678_2_edit <- is1678_2 %>% select(-c("IS1678"))
is1678_2_edit1 <- cbind(is1678_2_edit,is1678_2_values)

is1678_3 <- is_short %>% filter(IS1678 == "c(246241, 248570)" )
is1678_3_split <- as.data.frame(str_extract(string = is1678_3$IS1678,pattern = "(?<=\\().*(?=\\))") )
is1678_3_values <- as.data.frame(unlist(strsplit(is1678_3_split[1,1],','))) %>% t()
rownames(is1678_3_values) <- NULL
colnames(is1678_3_values) <- c("IS1678_2","IS1678")
is1678_3_edit <- is1678_3 %>% select(-c("IS1678"))
is1678_3_edit1 <- cbind(is1678_3_edit,is1678_3_values)

new_is1678 <- rbind(is1678_1_edit1,is1678_2_edit1,is1678_3_edit1)
new_islgar <- rbind(islgar_1_edit1,islgar_2_edit1,islgar_3_edit1)

no_is1678 <- anti_join(is_short,new_is1678,by="Clone")
no_either <- anti_join(no_is1678,new_islgar,by="Clone")


is_short2 <- merge(no_either,new_is1678, all=TRUE)
is_short3 <- merge(is_short2,new_islgar, all=TRUE)
is_short3[is_short3=='NULL'] = NA


is_short3 <- is_short3 %>% mutate(across(everything(),as.character))


#merge variants, low cov regions, IS elements
variant_low_cov_is <- merge(is_short3,variant_low_cov, by="Clone", all=TRUE)
variant_low_cov_is[variant_low_cov_is=="NULL"] <- NA


```



```{r assign a unique pairid to each distinct variation combination}
#identify variation combinations
variant_low_cov_is_distinct <- variant_low_cov_is %>% select(-(Clone)) %>% distinct()
#make a unique number identifier for each combination of variations
pairid <- seq(1, nrow(variant_low_cov_is_distinct),by=1) 
#assign identifier to variaitons
variant_low_cov_is_distinct$pairid <- pairid

#get all column names
variables <- variant_low_cov_is_distinct %>% select(-(pairid)) %>% colnames() 

matched_variant_low_cov_is <- merge(variant_low_cov_is,variant_low_cov_is_distinct, by=variables, all=TRUE)
num_variations <- table(matched_variant_low_cov_is$pairid) %>% as.data.frame()
matched_clone_pairid <- matched_variant_low_cov_is %>% select(c(Clone,pairid))

```


 
```{r edit after manual inspection of variations}

p99<- matched_variant_low_cov_is %>% filter(pairid ==99) %>% select_if(~ !any(is.na(.)))
p99_edit <- p99 %>% select(-c("247122-256268"))
p99_edit$"247121-256268" <- "247121-256268"

p117<- matched_variant_low_cov_is %>% filter(pairid ==117) %>% select_if(~ !any(is.na(.)))
p117_edit <- p117
p117_edit$"256197-256198" <- "256197-256198"

p2<- matched_variant_low_cov_is %>% filter(pairid ==2) %>% select_if(~ !any(is.na(.)))
p2_edit <- p2 %>% select(-c("247887-247894","256197-256198","251950","256196"))
p2_edit$"247121-256268" <- "247121-256268"


p3<- matched_variant_low_cov_is %>% filter(pairid ==3) %>% select_if(~ !any(is.na(.)))
p3_edit <- p3 %>% select(-contains("-"))
p3_edit$"247121-256268" <- "247121-256268"

p16<- matched_variant_low_cov_is %>% filter(pairid ==16) %>% select_if(~ !any(is.na(.)))
p16_edit <- p16 %>% select(-c("247121-247877","247886-256268"))
p16_edit$"247121-256268" <- "247121-256268"


p17<- matched_variant_low_cov_is %>% filter(pairid ==17) %>% select_if(~ !any(is.na(.)))
p17_edit <- p17 %>% select(-contains("-"))
p17_edit$"247121-256268" <- "247121-256268"

p19<- matched_variant_low_cov_is %>% filter(pairid ==19) %>% select_if(~ !any(is.na(.)))
p19_edit <- p19 %>% select(-contains("-"))
p19_edit$"247121-256268" <- "247121-256268"

p23<- matched_variant_low_cov_is %>% filter(pairid ==23) %>% select_if(~ !any(is.na(.)))
p23_edit <- p23 %>% select(-contains("-"))
p23_edit$"247121-256268" <- "247121-256268"

p27<- matched_variant_low_cov_is %>% filter(pairid ==27) %>% select_if(~ !any(is.na(.)))
p27_edit <- p27 %>% select(-contains("-"))
p27_edit$"246177-256268" <- "246177-256268"

p28<- matched_variant_low_cov_is %>% filter(pairid ==28) %>% select_if(~ !any(is.na(.)))
p28_edit <- p28 %>% select(-c("249774-249797","249798-249803"))

p30<- matched_variant_low_cov_is %>% filter(pairid ==30) %>% select_if(~ !any(is.na(.)))
p30_edit <- p30 %>% select(-contains("-"))
p30_edit$"247121-256268" <- "247121-256268"

p33<- matched_variant_low_cov_is %>% filter(pairid ==33) %>% select_if(~ !any(is.na(.)))
p33_edit <- p33 %>% select(-contains("-"))
p33_edit$"247121-256268" <- "247121-256268"

p34<- matched_variant_low_cov_is %>% filter(pairid ==34) %>% select_if(~ !any(is.na(.)))
p34_edit <- p34 %>% select(-contains("-"))
p34_edit$"246177-256268" <- "246177-256268"

p36<- matched_variant_low_cov_is %>% filter(pairid ==36) %>% select_if(~ !any(is.na(.)))
p36_edit <- p36 %>% select(-contains("-"))
p36_edit$"246177-256268" <- "246177-256268"

p37<- matched_variant_low_cov_is %>% filter(pairid ==37) %>% select_if(~ !any(is.na(.)))
p37_edit <- p37 %>% select(-contains("-"))
p37_edit$"247121-256268" <- "247121-256268"


p38<- matched_variant_low_cov_is %>% filter(pairid ==38) %>% select_if(~ !any(is.na(.)))
p38_edit <- p38 %>% select(-contains("-")) %>% select(-c("250694"))
p38_edit$"246177-256268" <- "246177-256268"

p40<- matched_variant_low_cov_is %>% filter(pairid ==40) %>% select_if(~ !any(is.na(.)))
p40_edit <- p40 %>% select(-contains("-"))
p40_edit$"246177-256268" <- "246177-256268"

p43<- matched_variant_low_cov_is %>% filter(pairid ==43) %>% select_if(~ !any(is.na(.)))
p43_edit <- p43 %>% select(-contains("-"))
p43_edit$"247121-256268" <- "247121-256268"

p44<- matched_variant_low_cov_is %>% filter(pairid ==44) %>% select_if(~ !any(is.na(.)))
p44_edit <- p44 %>% select(-contains("-"))
p44_edit$"247121-256268" <- "247121-256268"

p45<- matched_variant_low_cov_is %>% filter(pairid ==45) %>% select_if(~ !any(is.na(.)))
p45_edit <- p45 %>% select(-contains("-"))
p45_edit$"247121-256268" <- "247121-256268"

p46<- matched_variant_low_cov_is %>% filter(pairid ==46) %>% select_if(~ !any(is.na(.)))
p46_edit <- p46 %>% select(-contains("-"))
p46_edit$"247121-256268" <- "247121-256268"

p47<- matched_variant_low_cov_is %>% filter(pairid ==47) %>% select_if(~ !any(is.na(.)))
p47_edit <- p47 %>% select(-contains("-"))
p47_edit$"247121-256268" <- "247121-256268"

p63<- matched_variant_low_cov_is %>% filter(pairid ==63) %>% select_if(~ !any(is.na(.)))
p63_edit <- p63 %>% select(-contains("-")) %>% select(-c("IS1678","251950","256196"))
p63_edit$"247121-256268" <- "247121-256268"

p39<- matched_variant_low_cov_is %>% filter(pairid ==39) %>% select_if(~ !any(is.na(.)))
p39_edit <- p39 %>% select(-contains("-")) 
p39_edit$"247121-256268" <- "247121-256268"

p78<- matched_variant_low_cov_is %>% filter(pairid ==78) %>% select_if(~ !any(is.na(.)))
p78_edit <- p78 %>% select(-c("246683-246696"))

p79<- matched_variant_low_cov_is %>% filter(pairid ==79) %>% select_if(~ !any(is.na(.)))
p79_edit <- p79 %>% select(-c("246702-246831"))

p82<- matched_variant_low_cov_is %>% filter(pairid ==82) %>% select_if(~ !any(is.na(.)))
p82_edit <- p82 %>% select(-contains("-")) %>% select(-c("246717"))
p82_edit$"246655-248570" <- "246655-248570"


p85<- matched_variant_low_cov_is %>% filter(pairid ==85) %>% select_if(~ !any(is.na(.)))
p85_edit <- p85 %>% select(-c("246800-246811"))

p95<- matched_variant_low_cov_is %>% filter(pairid ==95) %>% select_if(~ !any(is.na(.)))
p95_edit <- p95 %>% select(-contains("-")) 
p95_edit$"247939-248084" <- "247939-248084"

p96<- matched_variant_low_cov_is %>% filter(pairid ==96) %>% select_if(~ !any(is.na(.)))
p96_edit <- p96 %>% select(-contains("-")) 
p96_edit$"246453-256268" <- "246453-256268"

p108<- matched_variant_low_cov_is %>% filter(pairid ==108) %>% select_if(~ !any(is.na(.)))
p108_edit <- p108 %>% select(-c("246398-246563"))

p109<- matched_variant_low_cov_is %>% filter(pairid ==109) %>% select_if(~ !any(is.na(.)))
p109_edit <- p109 %>% select(-contains("-")) 
p109_edit$"246342-246372" <- "246342-246372"
p109_edit$"246590-256268" <- "246590-256268"

p110<- matched_variant_low_cov_is %>% filter(pairid ==110) %>% select_if(~ !any(is.na(.)))
p110_edit <- p110 %>% select(-contains("-")) 
p110_edit$"246304-256268" <- "246304-256268"

p113<- matched_variant_low_cov_is %>% filter(pairid ==113) %>% select_if(~ !any(is.na(.)))
p113_edit <- p113 %>% select(-c("246831-246836","246867-246903"))

p114<- matched_variant_low_cov_is %>% filter(pairid ==114) %>% select_if(~ !any(is.na(.)))
p114_edit <- p114 %>% select(-contains("-")) 
p114_edit$"247121-256268" <- "247121-256268"

p119<- matched_variant_low_cov_is %>% filter(pairid ==119) %>% select_if(~ !any(is.na(.)))
p119_edit <- p119 %>% select(-c("1-246633"))
p119_edit$"246177-246633" <- "246177-246633"


p121<- matched_variant_low_cov_is %>% filter(pairid ==121) %>% select_if(~ !any(is.na(.)))
p121_edit <- p121 %>% select(-c("246673-246711"))

p122<- matched_variant_low_cov_is %>% filter(pairid ==122) %>% select_if(~ !any(is.na(.)))
p122_edit <- p122 %>% select(-contains("-")) 
p122_edit$"246484-256268" <- "246484-256268"

p123<- matched_variant_low_cov_is %>% filter(pairid ==123) %>% select_if(~ !any(is.na(.)))
p123_edit <- p123 %>% select(-c("246692-246773"))


edited_pairids_list <- list(p99_edit,p117_edit,p2_edit,p3_edit,p16_edit,
                            p17_edit,p19_edit,p23_edit,p27_edit,p28_edit,p30_edit,p34_edit,p36_edit,
                            p37_edit,p38_edit,p40_edit,p43_edit,p44_edit,p45_edit,p46_edit,p47_edit,
                            p63_edit,p39_edit,p78_edit,p79_edit,p82_edit,p85_edit,p95_edit,p96_edit,
                            p108_edit,p109_edit,p110_edit,p113_edit,p114_edit,p119_edit,p121_edit,
                            p122_edit,p123_edit)

#merge the new edited pairid variations together
edited_pairids <- Reduce(function(x, y) merge(x, y, all=TRUE), edited_pairids_list)
edited_pairids1 <- edited_pairids %>% select(-c("pairid"))

#edit the master variant file to remove the old pairid variants
matched_variant_low_cov_is1 <- matched_variant_low_cov_is %>% filter(!(pairid==99 | pairid==117 |
                                                                    pairid==2 |  pairid==3 |
                                                                    pairid==16 | pairid==17 |
                                                                    pairid==19 | pairid==23 |
                                                                    pairid==27 | pairid==28 |
                                                                    pairid==30 | pairid==34 |
                                                                    pairid==36 | pairid==37 |
                                                                    pairid==38 | pairid==40 |
                                                                    pairid==43 | pairid==44 |
                                                                    pairid==45 | pairid==46 |
                                                                    pairid==47 | pairid==63 |
                                                                    pairid==39 | pairid==78 |
                                                                    pairid==79 | pairid==82 |
                                                                    pairid==85 | pairid==95 |
                                                                    pairid==96 | pairid==108 |
                                                                    pairid==109 | pairid==110 |
                                                                    pairid==113 | pairid==114 |
                                                                    pairid==119 | pairid==121 |
                                                                    pairid==122 | pairid==123))
matched_variant_low_cov_is2 <- matched_variant_low_cov_is1 %>% select(-c("pairid"))

#merge the edited pairids with the master file 
matched_variant_low_cov_is_edited <- merge(edited_pairids1,matched_variant_low_cov_is2,all=TRUE)

```

```{r remove variants in low cov regions}
low_cov_new <- matched_variant_low_cov_is_edited %>% select("Clone",contains("-"))
lc <- low_cov_new %>% pivot_longer(!Clone, names_to = "low_cov", values_to = "count") 
lc1 <- lc[!is.na(lc$count),] %>% select(c("Clone","low_cov"))
lc_separate <- lc1 %>% separate(low_cov, c('Start_Position', 'Stop_Position'),"-")
lc_separate$Start_Position <- as.numeric(lc_separate$Start_Position)
lc_separate$Stop_Position <- as.numeric(lc_separate$Stop_Position)

snps_lc <- merge(annotated_accepted_variants ,lc_separate, by="Clone", all=TRUE)
snps_lc$between <- between(snps_lc$POS,snps_lc$Start_Position,snps_lc$Stop_Position)

snps_lc_btwn_true<- filter(snps_lc,between ==TRUE)

#remove snp calls that fall in low cov regions 
merged_snps_lc <- merge(annotated_accepted_variants,snps_lc_btwn_true, all.x=TRUE )
accepted_variants <- merged_snps_lc [is.na(merged_snps_lc$between), ] #only keep snp calls where the call in not in a low cov region
num_snps_accepted <- accepted_variants %>% select(c(POS,REF,ALT)) %>% distinct()

```




```{r merge variants, is, low cov}
variant_pivot <- pivot_wider(accepted_variants, id_cols="Clone", names_from= "POS", values_from="X.Annotation.Annotation_Impact.Gene_Name.Gene_ID.Feature_Type.")

merge2 <- merge(variant_pivot,low_cov_new,by="Clone",all=TRUE)
merge3<- merge(merge2, is_short3,by="Clone",all=TRUE)
```


```{r assign new pairid}
#assign a new pairid to the edited variations
#identify variation combinations
matched_variant_low_cov_is_edited_distinct <- merge3 %>% select(-(Clone)) %>% distinct()
#make a unique number identifier for each combination of variations
pairid <- seq(1, nrow(matched_variant_low_cov_is_edited_distinct),by=1) 
#assign identifier to variaitons
matched_variant_low_cov_is_edited_distinct$pairid <- pairid

#add pairids to each of the variations by clone 
matched_variant_low_cov_is_final <- merge(merge3,matched_variant_low_cov_is_edited_distinct, all=TRUE)
num_variations <- table(matched_variant_low_cov_is_final$pairid) %>% as.data.frame()
matched_clone_pairid <- matched_variant_low_cov_is_final %>% select(c(Clone,pairid))

```

```{r remove pairids with low cov across entire region}

matched_variant_low_cov_is_final1 <- matched_variant_low_cov_is_final
names(matched_variant_low_cov_is_final1)[names(matched_variant_low_cov_is_final1) == 'pairid'] <- 'pairid-246'
p<- matched_variant_low_cov_is_final1 %>% select(contains("-")) %>% select(contains("246"))

matched_variant_low_cov_is_final2 <- matched_variant_low_cov_is_final %>% filter(!(pairid==8|pairid==20|pairid==87|pairid==18|pairid==33))
num_variations1 <- table(matched_variant_low_cov_is_final2$pairid) %>% as.data.frame()


```





```{r final write for command line}
#write.table(x, args[5], row.names=FALSE, sep="\t",quote=FALSE)
write.table(matched_variant_low_cov_is_final2, "14_enterocinA_variation/summary/enterocinA_variation_summary_final.csv",row.names=FALSE, sep="\t",quote=FALSE)
```


```{r pairid39 matches 5}
data <- read.csv("14_enterocinA_variation/summary/enterocinA_variation_summary_final.csv",sep='\t')

p39 <- data %>% filter(pairid==39) %>%  select("Clone")
p5 <- data %>% filter(pairid==5) %>%  select(-c("Clone")) %>% distinct()
new_39 <- cbind(p39,p5)

data_filt <- data %>% filter(!(pairid==39))
data_final <- merge(data_filt,new_39,all=TRUE)
```

```{r pairid 62 missing low cov }
p62 <- data_final %>% filter(pairid==62)
p62$X256197.256198 <- "256197-256198"

data_final1 <- data_final %>% filter(!(pairid==62))
data_final2 <- merge(data_final1,p62,all=TRUE)

data_test <- data_final2 %>% select(-c("Clone","pairid")) %>% distinct()
write.table(data_final2, "14_enterocinA_variation/summary/enterocinA_variation_summary_final.csv",row.names=FALSE, sep="\t",quote=FALSE)

```

```{r pairid 40 matches 7}
p40 <- data_final2 %>% filter(pairid==40) %>%  select("Clone")
p7 <- data_final2 %>% filter(pairid==7) %>%  select(-c("Clone")) %>% distinct()
new_40 <- cbind(p40,p7)

data_filt <- data_final2 %>% filter(!(pairid==40))
data_final3 <- merge(data_filt,new_40,all=TRUE)

```

```{r pairid 31 matches 21}
p31 <- data_final3 %>% filter(pairid==31) %>%  select("Clone")
p21 <- data_final3 %>% filter(pairid==21) %>%  select(-c("Clone")) %>% distinct()
new_31 <- cbind(p31,p21)

data_filt <- data_final3 %>% filter(!(pairid==31))
data_final4 <- merge(data_filt,new_31,all=TRUE)
write.table(data_final4, "14_enterocinA_variation/summary/enterocinA_variation_summary_final.csv",row.names=FALSE, sep="\t",quote=FALSE)
data_test <- data_final4 %>% select(-c("Clone","pairid")) %>% distinct()

```


```{r format for phyloviz}
data <- read.csv("14_enterocinA_variation/summary/enterocinA_variation_summary_final.csv",sep='\t')
num_variations <- table(data$pairid) %>% as.data.frame()

pairid_list <- data %>% select(-c("Clone")) %>% distinct()
pairid_first <- pairid_list %>% select("pairid",everything())

p<- data %>% filter(pairid==26|pairid==19|pairid==15)  %>% remove_empty(which = "cols")

write.table(pairid_first, "14_enterocinA_variation/summary/enterocinA_variation_pairids_final.csv",row.names=FALSE, sep="\t",quote=FALSE)


mlst <- read.csv("/nfs/turbo/umms-robertwo/agarret/20230801_bacteriocin_blast/data/first_efm_pr_bl_mlst_dateQ.csv",sep='\t')
colnames(mlst) <- c("X","Clone","PID","cdate","ST","quarter")
clone_pairid_list <- data %>% select(c("Clone","pairid"))

clone_mlst_pairid <- merge(clone_pairid_list,mlst, by= "Clone")
clone_mlst_pairid1 <- clone_mlst_pairid %>% select(-c("X","cdate"))
write.table(clone_mlst_pairid1, "14_enterocinA_variation/summary/enterocinA_variation_clone_pairids_final.csv",row.names=FALSE, sep="\t",quote=FALSE)



```


```{r}
```{r choose isolates for phenotypic testing}
#pick one pairid of every major mlst (more than 20isolates in that st)
clone_mlst_pairid2 <- clone_mlst_pairid1 %>% select("Clone","pairid","ST") %>% group_by(pairid,ST)
clone_mlst <- clone_mlst_pairid2 %>% arrange(ST) %>% filter(row_number()==1) %>% arrange(pairid)

```

```
